Gems/Atom/Feature/Common/Assets/Materials/Types/DepthPass_WithPS.azsl

```diff
@@ -0,0 +1,124 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+struct VSInput
+{
+    float3 m_position : POSITION;
+    float2 m_uv0 : UV0;
+    float2 m_uv1 : UV1;
+
+    // only used for parallax depth calculation
+    float3 m_normal : NORMAL;
+    float4 m_tangent : TANGENT; 
+    float3 m_bitangent : BITANGENT; 
+
+#ifdef MULTILAYER
+    // This gets set automatically by the system at runtime only if it's available.
+    // There is a soft naming convention that associates this with o_blendMask_isBound, which will be set to true whenever m_optional_blendMask is available.
+    // (search "m_optional_" in ShaderVariantAssetBuilder for details on the naming convention).
+    // [GFX TODO][ATOM-14475]: Come up with a more elegant way to associate the isBound flag with the input stream.
+    float4 m_optional_blendMask : COLOR0;
+#endif
+};
+ 
+struct VSDepthOutput
+{
+    // "centroid" is needed for SV_Depth to compile
+    precise linear centroid float4 m_position : SV_Position;
+    float2 m_uv[UvSetCount] : UV1;
+
+    // only used for parallax depth calculation
+    float3 m_normal : NORMAL;
+    float3 m_tangent : TANGENT; 
+    float3 m_bitangent : BITANGENT; 
+    float3 m_worldPosition : UV0;
+
+#ifdef MULTILAYER
+    float3 m_blendMask : UV3;
+#endif
+};
+
+VSDepthOutput MainVS(VSInput IN)
+{
+    VSDepthOutput OUT;
+ 
+    float4x4 objectToWorld = GetObjectToWorld();
+    float4 worldPosition = mul(objectToWorld, float4(IN.m_position, 1.0));
+
+    OUT.m_position = mul(ViewSrg::m_viewProjectionMatrix, worldPosition);
+
+    float2 uvs[UvSetCount] = { IN.m_uv0, IN.m_uv1 };
+    TransformUvs(uvs, OUT.m_uv);
+
+    if(ShouldHandleParallaxInDepthShaders())
+    {
+        OUT.m_worldPosition = worldPosition.xyz;
+
+        float3x3 objectToWorldIT = GetNormalToWorld();
+        ConstructTBN(IN.m_normal, IN.m_tangent, IN.m_bitangent, objectToWorld, objectToWorldIT, OUT.m_normal, OUT.m_tangent, OUT.m_bitangent);
+    }
+
+#ifdef MULTILAYER
+    if(o_blendMask_isBound)
+    {
+        OUT.m_blendMask = IN.m_optional_blendMask.rgb;
+    }
+    else
+    {
+        OUT.m_blendMask = float3(0,0,0);
+    }
+#endif
+
+    return OUT;
+}
+
+struct PSDepthOutput
+{
+    precise float m_depth : SV_Depth;
+};
+
+PSDepthOutput MainPS(VSDepthOutput IN, bool isFrontFace : SV_IsFrontFace)
+{
+    PSDepthOutput OUT;
+
+    OUT.m_depth = IN.m_position.z;
+
+    if(ShouldHandleParallaxInDepthShaders())
+    {
+        float3 tangents[UvSetCount] = { IN.m_tangent, IN.m_tangent };
+        float3 bitangents[UvSetCount] = { IN.m_bitangent, IN.m_bitangent };
+
+        for (int i = 0; i != UvSetCount; ++i)
+        {
+            EvaluateTangentFrame(
+                IN.m_normal,
+                IN.m_worldPosition, 
+                isFrontFace,
+                IN.m_uv[i],
+                i,
+                IN.m_tangent,
+                IN.m_bitangent,
+                tangents[i],
+                bitangents[i]);
+        }
+
+#ifdef MULTILAYER
```

Comment:
Is this a direction we want to take? it makes the design even more 'uber-shader'.

Reply:
Yes. This single define resulted in a deduplication of hundreds of lines of code, which is strictly less maintainable and also results in the same number of shaders being compiled anyways.